src/runtime_safe_heap.js

```
@@ -8,91 +8,39 @@
 #error "should only be inclded in SAFE_HEAP mode"
 #endif
 
-/** @param {number|boolean=} isFloat */
-function getSafeHeapType(bytes, isFloat) {
-  switch (bytes) {
-    case 1: return 'i8';
-    case 2: return 'i16';
-    case 4: return isFloat ? 'float' : 'i32';
-    case 8: return isFloat ? 'double' : 'i64';
-    default: abort(`getSafeHeapType() invalid bytes=${bytes}`);
-  }
-}
-
 #if SAFE_HEAP_LOG
 var SAFE_HEAP_COUNTER = 0;
 #endif
 
-/** @param {number|boolean=} isFloat */
-function SAFE_HEAP_STORE(dest, value, bytes, isFloat) {
+function SAFE_HEAP_INDEX(arr, idx, action) {
 #if CAN_ADDRESS_2GB
-  dest >>>= 0;
+  idx >>>= 0;
 #endif
+  const bytes = arr.BYTES_PER_ELEMENT;
+  const dest = idx * bytes;
 #if SAFE_HEAP_LOG
-  dbg('SAFE_HEAP store: ' + [dest, value, bytes, isFloat, SAFE_HEAP_COUNTER++]);
-#endif
-  if (dest <= 0) abort(`segmentation fault storing ${bytes} bytes to address ${dest}`);
-#if SAFE_HEAP == 1
-  if (dest % bytes !== 0) abort(`alignment error storing to address ${dest}, which was expected to be aligned to a multiple of ${bytes}`);
-#else
-  if (dest % bytes !== 0) warnOnce(`alignment error in a memory store operation, alignment was a multiple of ${(((dest ^ (dest-1)) >> 1) + 1)}, but was was expected to be aligned to a multiple of ${bytes}`);
```

What happened to the alignment checks?

---

No longer necessary because we pass in an index in the corresponding array instead of pre-multiplied pointer.

---

Good point. I was also concerned about directly-emitted `SAFE_HEAP_*` calls from parseTools (not from the acorn transform), but I see we removed those a while ago, which I forgot...

---

I agree, yeah, best to just remove that bit, good find.

---

Hm this section https://emscripten.org/docs/debugging/Sanitizers.html#comparison-to-safe-heap also seems outdated:

> The specific things SAFE_HEAP errors on include:
> 
> [...]
> 
> Unaligned reads or writes. These work in WebAssembly, but on some platforms an incorrectly-aligned read or write may be much slower, and with wasm2js (WASM=0) it will be incorrect, as JavaScript Typed Arrays do not allow unaligned operations.

Are unaligned reads/writes no longer supported on wasm2js or are they handled differently now?

---

No, that still seems accurate to me? Reading a 32-bit number from unaligned address 2 is `HEAP32[2 >> 2] = HEAP32[0]` which is wrong.

---

`0` seems like easy case, but where would something like `HEAP32[123 >> 2]` -> `HEAP32[30]` be caught and handled? It seems at the moment the remainder part of the division is always ignored?

---

Normally it is silently ignored and behaves wrongly, but SAFE_HEAP in wasm checks it here:

https://github.com/WebAssembly/binaryen/blob/23a9e69f12afd749f35308dc504c8f6a89fc75f1/src/passes/SafeHeap.cpp#L376-L391

---

I see. So, just to confirm, the comment above refers specifically to Wasm (and its wasm2js output) and it's not a bug that `SAFE_HEAP` JS transform doesn't perform a separate check for unaligned reads/writes?

---

No, I think it is a bug that SAFE_HEAP in JS doesn't do alignment checks. We should in principle check the same things in both places (say someone moves code from JS to wasm or back - we should check safety in the same way). But, I think we saw that that JS-side bug is long-standing and never caused any issues, so it doesn't feel urgent to fix.

---

If the input code contains `HEAP32[x]` then by construction `x` must be 4-byte aligned, no?

---

Yes, but what about this?
```js
// a JS function that is called from wasm, and given a (maybe unaligned) pointer
function jsHelper(ptr) {
  return HEAP32[ptr >> 2];
}
```

---

> But, I think we saw that that JS-side bug is long-standing and never caused any issues, so it doesn't feel urgent to fix.

Fair enough. I was mainly curious in context of those no-op checks we are commenting on, as I assume they used to do something useful in the long past, so technically a regression happened at some point... but idk when. As you say, it's probably not urgent.

> If the input code contains `HEAP32[x]` then by construction `x` must be 4-byte aligned, no?

Problem I'm thinking about, a lot of code has form like `HEAP32[somePtr >> 2]` and such, whether handcrafted or generated via `makeGetValue`, and in those cases there is no guarantee `somePtr` is already correctly aligned.

JS transform could detect at least this common form of access and check the alignment. Or maybe at least `makeGetValue`/`getHeapOffset` itself could insert those checks when `SAFE_HEAP` is enabled.

---

> Yes, but what about this?
> 
> ```js
> // a JS function that is called from wasm, and given a (maybe unaligned) pointer
> function jsHelper(ptr) {
>   return HEAP32[ptr >> 2];
> }
> ```

But I don't think the old code handled this did it?  It just checked the value of `ptr >> 2` at runtime, right?   So there is no regression here is here?

---

> JS transform could detect at least this common form of access and check the alignment. Or maybe at least `makeGetValue`/`getHeapOffset` itself could insert those checks when `SAFE_HEAP` is enabled.

To be clear neither the old new the version the JS transform do this though do they?

---

> To be clear neither the old new the version the JS transform do this though do they?

If you mean before/after this PR, then yeah.

But judging by the removed code, it used to in the long past. And, well, even if it didn't, just saying it would be a useful addition.